src/librustc_mir/transform/match_branches.rs

```diff
@@ -0,0 +1,102 @@
+use crate::transform::{simplify, MirPass, MirSource};
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+
+pub struct MatchBranchSimplification;
+
+// What's the intent of this pass?
+// If one block is found that switches between blocks which both go to the same place
+// AND both of these blocks set a similar const in their ->
+// condense into 1 block based on discriminant AND goto the destination afterwards
+
+impl<'tcx> MirPass<'tcx> for MatchBranchSimplification {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, src: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        let param_env = tcx.param_env(src.def_id());
+        let mut did_remove_blocks = false;
+        let bbs = body.basic_blocks_mut();
+        'outer: for bb_idx in bbs.indices() {
+            let (discr, val, switch_ty, targets) = match bbs[bb_idx].terminator().kind {
+                TerminatorKind::SwitchInt {
+                    discr: Operand::Move(ref place),
+                    switch_ty,
+                    ref targets,
+                    ref values,
+                    ..
+                } if targets.len() == 2 && values.len() == 1 => {
+                    (place.clone(), values[0], switch_ty, targets)
+                }
+                _ => continue,
+            };
+            let (first, rest) = if let ([first], rest) = targets.split_at(1) {
+                (*first, rest)
+            } else {
+                unreachable!();
+            };
+            let first_dest = bbs[first].terminator().kind.clone();
+            let same_destinations = rest
+                .iter()
+                .map(|target| &bbs[*target].terminator().kind)
+                .all(|t_kind| t_kind == &first_dest);
+            if !same_destinations {
+                continue;
+            }
+            let first_stmts = &bbs[first].statements;
+            for s in first_stmts.iter() {
+                match &s.kind {
+                    StatementKind::Assign(box (_, rhs)) => {
+                        if let Rvalue::Use(Operand::Constant(_)) = rhs {
+                        } else {
+                            continue 'outer;
+                        }
+                    }
+                    _ => continue 'outer,
+                }
+            }
```

Comment:
the statements of `first` and `rest` should match up. maybe combine the `bbs[*target].statements loop below with the one above with `zip` (and first checking that they have the same length), and for every statement that is not the assignment you are looking for, ensure that they are the same statement

Reply:
Oh we're ok with other statements?

Reply:
well, ideally we only accept a single statement and assume that there are no others, but I think right now you're accepting multiple statements?

Reply:
The code you wrote now lgtm and is a more general optimization, so let's roll with it!